v3/newrelic/code_level_metrics.go

```diff
@@ -33,15 +34,70 @@ type CodeLocation struct {
 	FilePath string
 }
 
+//
+// CachedCodeLocation provides storage for the code location computed such that
+// the discovery of the code location is only done once; thereafter the cached
+// value is available for use.
+//
+// This type includes methods with the same names as some of the basic code location
+// reporting functions and TraceOptions. However, when called as methods of a CachedCodeLocation value
+// instead of a stand-alone function, the operation will make use of the cache to
+// prevent computing the same source location more than once.
+//
+// A usable CachedCodeLocation value must be obtained via a call to
+// NewCachedCodeLocation.
+//
+type CachedCodeLocation struct {
+	location *CodeLocation
+	once     *sync.Once
```

Comment:
Consider using `atomic.LoadPointer`, etc. instead of `sync.Once`. While it won't be air-tight (two goroutines might do the same lookup in parallel) it should be good enough for what you need, and avoids the extra logic you had to add to omit sync from the stack frame lookup.

Reply:
Good suggestion, and I considered that earlier on, but I think it would be better to go to the effort to use the more robust `sync.Once` function to avoid possible parallel lookups and such, and I think the way the code now steps around the calls to whatever's in the middle (e.g. `sync.Once`) will be flexible enough to support future needs as well.

Reply:
I think we would rather be defensive here, but appreciate the suggestion